### ISSUE TITLE
[EPIC] Model Zoo Architecture - 110+ Models for All Baseball Predictions

### ISSUE THREAD

**Issue by cbwinslow** [cbwinslow/retrosheet#131]

**Comment by cbwinslow** [comment]
# Model Zoo Architecture - Comprehensive Prediction System

## Vision

> "If there's data to predict it, we should have a model for it."

Train 110+ models covering **every conceivable baseball prediction** across 8 abstraction layers. Brute-force approach: train many models, evaluate all, keep the best.

## The 8 Layers of Abstraction

```
L1: League/Season    → "Will HRs be up this year?"
L2: Team             → "Will Tigers win 90 games?"
L3: Game             → "Who wins today?"
L4: Inning           → "Runs this inning?"
L5: PA               → "K, BB, or hit?"
L6: Count            → "Next pitch type?"
L7: Pitch            → "Swing? Contact?"
L8: Player/Career    → "Harper's next milestone?"
```

## Model Count by Layer

| Layer | Count | Examples |
|-------|-------|----------|
| L1 League | 5 | Season trends, playoff odds |
| L2 Team | 10 | Win totals, streaks, form |
| L3 Game | 15 | Winner, spread, total |
| L4 Inning | 8 | Runs, batting around |
| L5 PA | 20 | HR, K, BB, outcome distribution |
| L6 Count | 15 | Pitch type, count progression |
| L7 Pitch | 12 | Swing, contact, fair, location |
| L8 Player | 25+ | Milestones, career trajectory, star-player specific |
| **Total** | **110+** | Every baseball prediction |

## Key Design Decisions

### 1. Player-Specific Models
- Create materialized views for star players
- If processing allows, extend to all players
- Generic models as fallback

### 2. All Prediction Horizons
- Next pitch
- Full PA outcome
- Rest of inning
- Full game
- Season/career

### 3. Model Selection
- User selects models (to be implemented)
- For now: train everything, evaluate later
- Registry tracks performance for selection

### 4. Training Data
- All historical data (2015-2024)
- Time-based splits (no shuffle)
- Regular retraining

### 5. Historical Patterns ("Announcer Stats")
- **Both** pre-computed AND real-time SQL
- Pattern database for common queries
- Live SQL for novel patterns

### 6. Latency
- **<5 seconds** acceptable
- Allows complex ensembles
- Materialized views for speed

### 7. Model Storage
- **Database-native** (MLflow-style)
- `models.registry` table with BYTEA blobs
- Versioning, metadata, performance tracking

## Model Registry Features

**Table: `models.registry`**
- Model binary (BYTEA)
- Performance metrics (accuracy, F1, RMSE)
- Overfitting protection (train/test gap)
- Hyperparameters
- Feature importance
- Status (training/active/deprecated)

**Functions:**
- `models.get_best_model()` - Select optimal model
- `models.log_prediction()` - Track live predictions
- `models.detect_drift()` - Monitor model degradation

## Overfitting Protection

1. **Time-based splits**: Train 2015-2021, Val 2022-2023, Test 2024
2. **Gap monitoring**: Alert if train_accuracy - test_accuracy > 5%
3. **Regularization**: XGBoost max_depth=6, dropout=0.3 for NN
4. **Cross-validation**: 5-fold time-series CV
5. **Backtesting**: Run on 2020-2023, compare to actual

## Natural Language Query Examples

```python
# Example queries the system should handle
"Bottom 9th, 2-2 count, Scherzer vs Harper - HR chance?"
→ L5: pa_is_home_run, L3: game_winner (context)

"When's the last time a Tiger under 25 hit back-to-back doubles?"
→ L8: historical_pattern (pre-computed + live SQL)

"What's Harper's chance of 600 career HRs?"
→ L8: career_milestone_probability (player-specific)

"Will this pitch be a fastball?"
→ L6: count_pitch_type_classifier

"Swing probability on this pitch?"
→ L7: pitch_swing_probability
```

## Deliverables

### Phase 1: Infrastructure (This Session)
- [x] Model Zoo Architecture document
- [x] Model Registry SQL schema
- [ ] Training pipeline framework
- [ ] Model serving layer
- [ ] NL query router

### Phase 2: First Models
- [ ] Pitch-level: swing, contact, type
- [ ] PA-level: outcome distribution
- [ ] Game-level: Monte Carlo simulation
- [ ] Player-specific: Harper, Ohtani, Judge

### Phase 3: Historical Patterns
- [ ] Pattern database schema
- [ ] Pre-computation pipeline
- [ ] Live SQL query generator

### Phase 4: Production
- [ ] Model auto-retraining
- [ ] Drift detection
- [ ] Performance monitoring dashboard

## Documents

- `docs/MODEL_ZOO_ARCHITECTURE.md` - Full architecture
- `sql/60_models/6006_model_registry_schema.sql` - Database schema
- `docs/HPS_DATA_DICTIONARY.md` - 130+ features available

## Next Steps

1. Implement training pipeline Python framework
2. Create first batch of models (pitch + PA)
3. Build NL query router
4. Set up materialized views for star players
